src/debug.cc

```diff
@@ -92,6 +459,219 @@ ResyncDebugLog(FILE *newFile)
+    // used. We now know that it will be used. Also logs saved early messages
+    // (if they became eligible now) before lastWrittenRecordNumber blocks them.
+    stopEarlyMessageCollection();
+
+    write(header, body);
+}
+
+void
+Debug::ForgetSaved()
+{
+    auto &module = Module();
+    (void)module.cacheLogChannel.releaseEarlyMessages();
+    (void)module.stderrChannel.releaseEarlyMessages();
+    (void)module.syslogChannel.releaseEarlyMessages();
+}
+
+void
+Debug::PrepareToDie()
+{
+    Module().prepareToDie();
+}
+
+void
+DebugChannel::StopSavingAndLog(DebugChannel &channelA, DebugChannel *channelBOrNil)
+{
+    const LoggingSectionGuard sectionGuard;
+
+    assert(&channelA != channelBOrNil);
+    const auto asOrNil = channelA.releaseEarlyMessages();
+    const auto bsOrNil = channelBOrNil ? channelBOrNil->releaseEarlyMessages() : nullptr;
+    const auto &as = asOrNil ? *asOrNil : CompiledDebugMessages();
+    const auto &bs = bsOrNil ? *bsOrNil : CompiledDebugMessages();
+
+    const auto writtenEarlier = channelA.written;
+
+    std::merge(as.begin(), as.end(), bs.begin(), bs.end(), Logger(channelA),
+    [](const CompiledDebugMessage &mA, const CompiledDebugMessage &mB) {
+        return mA.header.recordNumber < mB.header.recordNumber;
+    });
+
+    const auto writtenNow = channelA.written - writtenEarlier;
+    if (const auto totalCount = as.size() + bs.size()) {
+        debugs(0, 5, "wrote " << writtenNow << " out of " << totalCount << '=' <<
+               as.size() << '+' << bs.size() << " early messages to " << channelA.name);
+    }
+}
+
+void
+DebugChannel::saveMessage(const DebugMessageHeader &header, const CompiledDebugMessageBody &body)
+{
+    if (!earlyMessages)
+        return; // we have stopped saving early messages
+
+    if (header.level > EarlyMessagesLevel)
+        return; // this message is not important enough to save
+
+    // Given small EarlyMessagesLevel, only a Squid bug can cause so many
+    // earlyMessages. Saving/dumping excessive messages correctly is not only
+    // difficult but is more likely to complicate triage than help: It is the
+    // first earlyMessages that are going to be the most valuable. Our assert()
+    // will dump them if at all possible.
+    assert(earlyMessages->size() < 1000);
+
+    earlyMessages->emplace_back(header, body);
+}
+
+void
+DebugChannel::writeToStream(FILE &destination, const DebugMessageHeader &header, const CompiledDebugMessageBody &body)
+{
+    fprintf(&destination, "%s%s| %s\n",
+            debugLogTime(header.timestamp),
+            debugLogKid(),
+            body.c_str());
+    noteWritten(header);
+}
+
+void
+DebugChannel::noteWritten(const DebugMessageHeader &header)
+{
+    ++written;
+    lastWrittenRecordNumber = header.recordNumber;
+}
+
+/* CacheLogChannel */
+
+bool
+CacheLogChannel::shouldWrite(const DebugMessageHeader &) const
+{
+    return TheLog.file();
+}
```

Comment:
Please inline this and similar one-liner methods especially if they are virtuals.

Reply:
Why? I think this method should _not_ be inlined especially because it is virtual:

* all similar methods are defined consistently this way (because most are not one-liners);
* inlining distracts the reader from comprehending the class declaration
* minor code changes (e.g., adding a source code comment) will not require moving the method;
* inlining of virtual methods is unlikely, but the optimizing compiler will inline (if it can) regardless of where we put the definition;
* method definitions (and their history of changes) are easier to find when they are _not_ inlined

Reply:
> Why?

Primarily as an optimization to avoid any method jump or virtual overheads in the logic which can be inlined. A bunch of the "thousand cuts" against Squid performance is coming from simple methods like this which could be inlined but are not.

> I think this method should _not_ be inlined especially because it is virtual:
> 
>     * all similar methods are defined consistently this way (because most are not one-liners);

Only if you isolate your view to this class. The wider Squid code style regularly inlines simple methods like these.

>     * inlining distracts the reader from comprehending the class declaration

Not important. Verbosity of the mandatory comments does more harm there than one-liner implementations. In fact the implementation being present can help avoid some of the comment.

>     * minor code changes (e.g., adding a source code comment) will not require moving the method;
 
Moving is not difficult, and with our emphasis on design perfection should be a rarity for methods this simple.

FYI, I am okay with inline methods having a few lines even. I just dont request it for 2+ and leave it to the authors judgment on whether such a move just for a comment is worth it.

>     * inlining of virtual methods is unlikely, but the optimizing compiler will inline (if it can) regardless of where we put the definition;

Only advanced optimization levels are capable of that when the definition is in the .cc file. The compileler may not even have access to the on-line definition at simpler optimization levels. By placing it in the header we guarantee that even un-optimized builds can inline and gain the performance benefits.

> 
>     * method definitions (and their history of changes) are easier to find when they are _not_ inlined

IME this is not true, history is easily available with git regardless of where they are placed.

Reply:
> > Alex: Why?
> Amos: Primarily as an optimization to avoid any method jump or virtual overheads in the logic which can be inlined.

In this case, a modern optimizing compiler will (not) inline regardless of where the method is defined. Also, the compiler will either know the exact address of the virtual function in the caller context or not, regardless of where the virtual method is defined. The definition _location_ has no affect on that knowledge (and, hence, on the compiler ability to create code that avoids runtime vtable lookups).

Your primary motivation is based on false assumptions. The secondary arguments below do not invalidate my points in defense of the current method definition location. Please withdraw your change request.


> > Alex: I think this method should _not_ be inlined especially because it is virtual:
> > * all similar methods are defined consistently this way (because most are not one-liners);
 
> Amos: Only if you isolate your view to this class. The wider Squid code style regularly inlines simple methods like these.

The wider Squid code that inlines virtual methods in similar circumstances is wrong. There is no one simple "inline all one-liners" rule. The decision should take more factors into account, and that is exactly what this PR is doing.

> > Alex: * inlining distracts the reader from comprehending the class declaration

> Amos: Not important.

It is important enough to be a factor IMO.

> Amos: Verbosity of the mandatory comments does more harm there than one-liner implementations. In fact the implementation being present can help avoid some of the comment.

Let's not discuss your misinterpretation of description rules and your opinion on this PR comments in _this_ change request. If you strongly believe some PR-added comments should be removed, please open dedicated change requests.

> > Alex: * minor code changes (e.g., adding a source code comment) will not require moving the method;

> Amos: Moving is not difficult, and with our emphasis on design perfection should be a rarity for methods this simple.

Your arguments do not invalidate my point, so I will not detail their serious flaws.

> > Alex: * inlining of virtual methods is unlikely, but the optimizing compiler will inline (if it can) regardless of where we put the definition;

> Amos: Only advanced optimization levels are capable of that when the definition is in the .cc file. The compileler may not even have access to the on-line definition at simpler optimization levels. By placing it in the header we guarantee that even un-optimized builds can inline and gain the performance benefits.

In this particular case, the definition and the declaration are both in the same `.cc` file so the compiler can see everything and safely assume a lot. I do not know what you call "advanced optimization levels", but the basic inlining we are talking about here starts with `-O2`. I would not call that level "advanced". In fact, that is usually the minimum level optimized builds should use. Finally, unoptimized builds do not inline even if the definition is provided at the declaration time! In summary, the arguments against my point do not apply to the use case in question or are invalid.
 
> > Alex: * method definitions (and their history of changes) are easier to find when they are _not_ inlined
 
> Amos: IME this is not true, history is easily available with git regardless of where they are placed.

I am not talking about availability. I am talking about the ease of finding/locating method definitions.

If you do not believe me, try writing a `git grep` search expression that locates the NullaryMemFunT::doDial() method definition and only that method definition, without relying on a particular implementation of that method, its formatting, and any other details that the searching developer (or script!) may not know or that may change from one git commit to another. If that method were not inlined, the task would be trivial, of course: `git grep NullaryMemFunT::doDial\b`.

Increased search difficulty goes beyond `git grep`, of course. I am just using that command as the simplest example.